core/jvm/src/main/scala/scalaz/zio/lockfree/impls/RingBuffer.scala

```diff
@@ -0,0 +1,294 @@
+ * 3. a mechanism to ''publish'' changes to `buf` via volatile write
+ *    (can even be relaxed to ordered store).
+ * See comments in `offer`/`poll` methods for more details on `seq`.
+ *
+ * The benefit of using `seq` + `head`/`tail` counters is that there
+ * are no allocations during enqueue/dequeue and very little
+ * overhead. The downside is it doubles (on 64bit) or triples
+ * (compressed OOPs) the amount of memory needed for queue.
+ *
+ * Concurrent enqueues and concurrent dequeues are possible. However
+ * there is no ''helping'', so threads can delay other threads, and
+ * thus the queue doesn't provide full set of lock-free
+ * guarantees. In practice it's usually not a problem, since benefits
+ * are simplicity, zero GC pressure and speed.
+ *
+ * The real capacity of the queue is the next power of 2 of the
+ * `desiredCapacity`. The reason is `head % N` and `tail % N` are
+ * rather cheap when can be done as a simple mask (N is pow 2), and
+ * pretty expensive when involve an `idiv` instruction. The queue can
+ * be made to work with arbitrary sizes but the user will have to
+ * suffer ~20% performance loss.
+ *
+ * The design is heavily inspired by such libraries as
+ * [[https://github.com/LMAX-Exchange/disruptor]] and
+ * [[https://github.com/JCTools/JCTools]] which is based off
+ * D. Vyukov's design
+ * [[http://www.1024cores.net/home/lock-free-algorithms/queues/bounded-mpmc-queue]]
+ *
+ * Compared to JCTools this implementation doesn't rely on
+ * [[sun.misc.Unsafe]], so it is arguably more portable, and should be
+ * easier to read. It's also very extensively commented, including
+ * reasoning, assumptions, and hacks.
+ *
+ * =Alternative designs=
+ *
+ * There is an alternative design described in
+ * [[http://pirkelbauer.com/papers/icapp16.pdf the paper]] A Portable
+ * Lock-Free Bounded Queue by Pirkelbauer et al.
+ *
+ * It provides full lock-free guarantees, which generally means that
+ * one out of many contending threads is guaranteed to make progress
+ * in a finite number of steps. The design thus is not susceptible to
+ * threads delaying other threads.
+ *
+ * However the helping scheme is rather involved and cannot be
+ * implemented without allocations (at least I couldn't come up with
+ * a way yet). This translates into worse performance on average, and
+ * better performance in some very specific situations.
+ */
+class RingBuffer[A](val desiredCapacity: Int) extends MutableConcurrentQueue[A] {
+  final val capacity: Int   = nextPow2(desiredCapacity)
+  private val idxMask: Long = (capacity - 1).toLong
+
+  private val buf: Array[AnyRef]   = new Array[AnyRef](capacity)
+  private val seq: AtomicLongArray = new AtomicLongArray(capacity)
+  0.until(capacity).foreach(i => seq.set(i, i.toLong))
+
+  private val head: AtomicLong = new AtomicLong(0L)
+  private val tail: AtomicLong = new AtomicLong(0L)
+
+  private final val STATE_LOOP     = 0
+  private final val STATE_EMPTY    = -1
```

Comment:
`private[this]`

Reply:
Sure! BTW, these constants get inlined by C2 so method access is not a problem. I haven't checked C1 assembly, though, `private[this]` could be beneficial in this case.